Lib/inspect.py

```diff
@@ -3355,11 +3397,135 @@ def format(self, *, max_width=None):
         return rendered
 
 
+class MultiSignature(Signature):
+    __slots__ = ('_signatures',)
+
+    def __init__(self, signatures):
+        signatures = tuple(signatures)
+        if not signatures:
+            raise ValueError('No signatures')
+        self._signatures = signatures
+
+    @staticmethod
+    def from_callable(obj, *,
+                      follow_wrapped=True, globals=None, locals=None, eval_str=False):
+        """Constructs MultiSignature for the given callable object."""
+        signature = Signature.from_callable(obj, follow_wrapped=follow_wrapped,
+                                            globals=globals, locals=locals,
+                                            eval_str=eval_str)
+        if not isinstance(signature, MultiSignature):
+            signature = MultiSignature((signature,))
+        return signature
+
+    @property
+    def parameters(self):
+        try:
+            return self._parameters
+        except AttributeError:
+            pass
+        params = {}
+        for s in self._signatures:
+            params.update(s.parameters)
+        self._parameters = types.MappingProxyType(params)
+        return self._parameters
+
+    @property
+    def return_annotation(self):
+        try:
+            return self._return_annotation
+        except AttributeError:
+            pass
+        return_annotations = []
+        for s in self._signatures:
+            ann = s.return_annotation
+            if ann != _empty:
+                if ann is None:
+                    ann = type(ann)
+                elif isinstance(ann, str):
+                    from typing import ForwardRef
+                    ann = ForwardRef(ann)
+                return_annotations.append(ann)
+        self._return_annotation = (functools.reduce(or_, return_annotations)
+                                   if return_annotations else _empty)
+        return self._return_annotation
+
+    def replace(self):
+        raise NotImplementedError
+
+    __replace__ = replace
+
+    def __iter__(self):
+        return iter(self._signatures)
+
+    def __hash__(self):
+        if len(self._signatures) == 1:
+            return hash(self._signatures[0])
+        return hash(self._signatures)
+
+    def __eq__(self, other):
+        if self is other:
+            return True
+        if isinstance(other, MultiSignature):
+            return self._signatures == other._signatures
+        if len(self._signatures) == 1 and isinstance(other, Signature):
```

Comment:
I think this should be 
```suggestion
        if isinstance(other, Signature):
            return len(self._signatures) == 1
```
and the addition to Signature's eq should be removed, above.

Reply:
No, it is not correct, it would make a 1-element MultiSignature equal to any Signature.

You perhaps mistyped, I wrote more correct fix.

Reply:
I did. That's good. (I would suggest removing the line about MultiSignature in Signature's eq as I believe it's dead code now, but up to you).

Reply:
It is not dead. It is called in `SignatureSubclass() == MultiSignature()`. Although it is not covered by tests yet.